watertap/unit_models/crystallizer.py

```diff
@@ -0,0 +1,696 @@
+        def eq_enthalpy_balance(b):
+            return (b.properties_in[0].enth_flow 
+                - b.properties_out[0].enth_flow - b.properties_vapor[0].enth_flow - b.properties_solids[0].enth_flow
+                + self.work_mechanical[0]
+                - (b.properties_solids[0].flow_vol_phase['Sol'] * b.properties_solids[0].dens_mass_solute['Sol'] * b.properties_solids[0].dh_crystallization['NaCl'])
+                == 0
+                )
+
+
+        # 6. Pressure and temperature balances - what is the pressure of the outlet solid and vapour?
+        # TO-DO: Figure out actual liquid and solid pressures.
+        @self.Constraint()    
+        def eq_p_con1(b):
+            return b.properties_in[0].pressure - b.properties_out[0].pressure == 0
+
+        @self.Constraint()    
+        def eq_p_con2(b):
+            return b.properties_in[0].pressure - b.properties_solids[0].pressure == 0
+
+        @self.Constraint()    
+        def eq_p_con3(b):
+            return b.properties_vapor[0].pressure - self.pressure_operating == 0
+
+        @self.Constraint()    
+        def eq_T_con1(b):
+            return self.temperature_operating - b.properties_solids[0].temperature == 0
+
+        @self.Constraint()    
+        def eq_T_con2(b):
+            return self.temperature_operating - b.properties_vapor[0].temperature == 0
+
+        @self.Constraint()
+        def eq_T_con3(b):
+            return self.temperature_operating - b.properties_out[0].temperature == 0
+
+
+        # 7. Heat exchanger minimum circulation flow rate calculations - see Lewis et al. or Tavare et al.
+        @self.Constraint(doc="Constraint on mimimum circulation rate through crystallizer heat exchanger")
+        def eq_minimum_hex_circulation_rate_constraint(b):
+            rho_cp_dt = self.hex_max_temp_increase * \
+            (
+                b.product_volumetric_solids_fraction * b.properties_solids[0].dens_mass_solute['Sol'] * b.properties_solids[0].cp_solute['Sol'] 
+                + (1 - b.product_volumetric_solids_fraction) * b.properties_out[0].dens_mass_phase['Liq'] * b.properties_out[0].cp_phase['Liq']
+                )
+            return (
+                b.magma_circulation_flow_vol * rho_cp_dt == pyunits.convert(b.work_mechanical[0], to_units=pyunits.J / pyunits.s)
+                )
+
+        # 8. Suspension density
+        @self.Constraint(doc="Slurry density calculation")
+        def eq_slurry_density(b):
+            return (self.slurry_density == 
+                b.product_volumetric_solids_fraction * b.properties_solids[0].dens_mass_solute['Sol'] 
+                + (1 - b.product_volumetric_solids_fraction) * b.properties_out[0].dens_mass_phase['Liq']
+                )
+
+
+        # 9. Residence time calculation
+        @self.Constraint(doc="Residence time")
+        def eq_residence_time(b):
+            return (
+                b.t_res == b.L_d / (3.67 * pyunits.convert(b.G_rate, to_units=pyunits.m/pyunits.hr))
```

Comment:
What is the 3.67? We should probably make that a parameter or variable for the unit model.

Reply:
It is not a variable. It is a constant in the equation relating residence time, crystal growth rate and crystal length:

crystal length L = 3.67 * Growth rate (m/s)  * residence time (s)

Reply:
3.67 moved ito a parameter called ``shape_factor_constant``